### PR TITLE
Use sharp to change screenshot size or convert to JPG [Merge for 6.1]

### PR DESCRIPTION
**Pull request by soulgalore**
We need to update the Dockerfile to include build-essential (higher up in the Docker chain).

### CODE REVIEW

File: lib/plugins/screenshot/index.js [hunk 1]
```diff
@@ -1,46 +1,98 @@
...
-      screenshot.name,
-      url,
-      'screenshots'
-    )
-  );
+function storeScreenshots(url, imagesAndName, storageManager, options) {
+  return Promise.map(imagesAndName, screenshot => {
+    const maxSize = get(options, 'maxSize', defaultConfig.maxSize);
+    const imageType = get(options, 'type', defaultConfig.type);
+    if (imageType === 'png') {
```
Comment (tobli) [inline lib/plugins/screenshot/index.js]: Do this if once, outside the map. Then just set an outputOptions object, to be passed to sharp.

File: lib/plugins/screenshot/index.js [hunk 2]
```diff
@@ -1,46 +1,98 @@
...
+      const compressionLevel = get(
+        options,
+        'png.compressionLevel',
+        defaultConfig.png.compressionLevel
+      );
+      sharp(screenshot.data)
+        .png({ compressionLevel })
```
Comment (tobli) [inline lib/plugins/screenshot/index.js]: Use [toFormat(imageType, outputOptions)](http://sharp.dimens.io/en/stable/api-output/#toformat) instead to remove the if/else.

File: lib/plugins/screenshot/index.js [hunk 3]
```diff
@@ -1,46 +1,98 @@
...
+        .resize(maxSize, maxSize)
+        .max()
+        .toBuffer()
+        .then(newBuff => {
```
Comment (tobli) [inline lib/plugins/screenshot/index.js]: this could be simplified to skip the { and return.

File: lib/plugins/browsertime/index.js [hunk 4]
```diff
@@ -271,10 +271,17 @@ module.exports = {
           .tap(results => {
             if (results.screenshots) {
               queue.postMessage(
-                make('browsertime.screenshot', results.screenshots, {
-                  url,
-                  group
-                })
+                make(
+                  'browsertime.screenshot',
+                  {
```
Comment (tobli) [inline lib/plugins/browsertime/index.js]: This is technically a backwards incompatible change (external plugins would need to change as we do). I think it's ok, just need to include a mention in the changelog (in the PR would be good).

A good reminder as well to audit all other plugins to always use a top level object for message data, to allow for further expansion without similar api breakage.

File: lib/plugins/html/templates/url/index.pug [hunk 5]
```diff
@@ -96,8 +96,9 @@ block content
         .one-half.column
             if hasScreenShots
               - var width = options.mobile ? 150 : '100%';
-              a(href='data/screenshots/0.png')
-                img.screenshot(src='data/screenshots/0.png', width=width, alt='Screenshot of run 1')
+              - var screenshotName = 'data/screenshots/0.' + screenShotType
+              a(href=screenshotName)
+                img.screenshot(src=screenshotName , width=width, alt='Screenshot of run 1')
```
Comment (tobli) [inline lib/plugins/html/templates/url/index.pug]: remove the space after screenshotName here.

File: lib/plugins/browsertime/index.js (hunk 4 above)
Comment (soulgalore) [reply lib/plugins/browsertime/index.js]: hmm that's true we should do it in another way.

File: lib/plugins/screenshot/index.js [hunk 6]
```diff
@@ -1,46 +1,98 @@
...
+  return Promise.map(imagesAndName, screenshot => {
+    const maxSize = get(options, 'maxSize', defaultConfig.maxSize);
+    const imageType = get(options, 'type', defaultConfig.type);
+    if (imageType === 'png') {
+      const compressionLevel = get(
+        options,
+        'png.compressionLevel',
+        defaultConfig.png.compressionLevel
+      );
+      sharp(screenshot.data)
```
[the remark this answers is not in the document]
Comment (soulgalore) [reply lib/plugins/screenshot/index.js]: fixed

File: lib/plugins/screenshot/index.js [hunk 7]
```diff
@@ -1,46 +1,98 @@
...
+          return storageManager.writeDataForUrl(
+            newBuff,
+            screenshot.name,
+            url,
+            'screenshots'
+          );
+        });
+    } else {
+      const quality = get(options, 'jpg.quality', defaultConfig.jpg.quality);
+      sharp(screenshot.data)
```
[the remark this answers is not in the document]
Comment (soulgalore) [reply lib/plugins/screenshot/index.js]: fixed

File: lib/plugins/screenshot/index.js [hunk 8]
```diff
@@ -1,46 +1,98 @@
...
-      screenshot.data,
-      screenshot.name,
-      url,
-      'screenshots'
-    )
-  );
+function storeScreenshots(url, imagesAndName, storageManager, options) {
+  return Promise.map(imagesAndName, screenshot => {
+    const maxSize = get(options, 'maxSize', defaultConfig.maxSize);
+    const imageType = get(options, 'type', defaultConfig.type);
```
[the remark this answers is not in the document]
Comment (soulgalore) [reply lib/plugins/screenshot/index.js]: fixed

File: lib/plugins/screenshot/index.js (hunk 1 above)
Comment (soulgalore) [reply lib/plugins/screenshot/index.js]: fixed

File: lib/plugins/screenshot/index.js (hunk 3 above)
Comment (soulgalore) [reply lib/plugins/screenshot/index.js]: fixed

File: lib/plugins/screenshot/index.js [hunk 9]
```diff
@@ -1,46 +1,98 @@
...
+    }
+  });
 }
 
 module.exports = {
   open(context, options) {
     this.storageManager = context.storageManager;
     this.make = context.messageMaker('screenshot').make;
-    this.options = options;
+    this.options = options.screenshot;
```
[the remark this answers is not in the document]
Comment (soulgalore) [reply lib/plugins/screenshot/index.js]: fixed (added specific screenshot config).

File: lib/plugins/screenshot/index.js [hunk 10]
```diff
@@ -1,46 +1,98 @@
...
+    this.imageType = get(this.options, 'type', defaultConfig.type);
```
[the remark this answers is not in the document]
Comment (soulgalore) [reply lib/plugins/screenshot/index.js]: fixed.

**Comment by tobli** [comment]
👍 Looks good to go with the changes (I didn't actually test it myself). Not sure if the message data compatibility is a blocker.

**Comment by soulgalore** [comment]
We could change so it is the actual screenshot plugin that adds the types, hmm.

File: lib/plugins/browsertime/index.js (hunk 4 above)
Comment (soulgalore) [reply lib/plugins/browsertime/index.js]: I've changed it now so that HTML listen on the browsertime config.

**Comment by soulgalore** [comment]
This should be good to go now.

**Pull request merged**